### PR TITLE
[SPARK-40755][SQL] Migrate type check failures of number formatting onto error classes

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr aims to replaces TypeCheckFailure by DataTypeMismatch in type checks in the number formatting or parsing expressions, includes:
1. ToNumber (1):
https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala#L83
2. ToCharacter (1):
https://github.com/apache/spark/blob/1431975723d8df30a25b2333eddcfd0bb6c57677/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala#L227
3. ToNumberParser (1):
https://github.com/apache/spark/blob/5556cfc59aa97a3ad4ea0baacebe19859ec0bcb7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala#L262

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
1. Add new UT
2. Update existed UT
3. Pass GA

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -290,6 +290,46 @@
           "Null typed values cannot be used as arguments of <functionName>."
         ]
       },
+      "NUMBER_FORMAT_CURRENCY_MUST_BEFORE_DECIMAL" : {
+        "message" : [
+          "Currency characters must appear before any decimal point in the number format: '<numberFormat>'"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:295]: Let's quote the format as we do in other places:
```
toSQLValue(pattern, StringType)
```
see, for instance
https://github.com/apache/spark/blob/0205478b9d35d62450fd7c9ade520087fd2979a7/sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala#L1381

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -290,6 +290,46 @@
...
+        ]
+      },
+      "NUMBER_FORMAT_CURRENCY_MUST_BEFORE_DIGIT" : {
+        "message" : [
+          "Currency characters must appear before digits in the number format: '<numberFormat>'"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:300]: ditto: quoting of `numberFormat`

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -290,6 +290,46 @@
...
+        ]
+      },
+      "NUMBER_FORMAT_EMPTY" : {
+        "message" : [
+          "The format string cannot be empty"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:305]: nit: but for consistency w/ other error messages. Could you add a dot at the end.
```suggestion
          "The format string cannot be empty."
```

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -290,6 +290,46 @@
...
+        ]
+      },
+      "NUMBER_FORMAT_THOUSANDS_SEPARATOR_MUST_BEFORE_DECIMAL" : {
+        "message" : [
+          "Thousands separators (, or G) may not appear after the decimal point in the number format: '<numberFormat>'"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:310]: ditto: quoting + dot at the end.

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -290,6 +290,46 @@
...
+        ]
+      },
+      "NUMBER_FORMAT_THOUSANDS_SEPARATOR_MUST_HAVE_DIGIT_IN_BETWEEN" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:313]: The name is too long. Is it possible to make it slightly shorter?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 6]
```diff
@@ -1134,6 +1181,21 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     }
   }
 
+  test("ToNumber: fails analysis if numberFormat is not a constant") {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1184]: nit: is not foldable

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 7]
```diff
@@ -1022,76 +1022,123 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
         s"The format string should have been invalid: $format")
-      toCharResult match {
-        case TypeCheckResult.TypeCheckFailure(message) =>
-          assert(message.contains(expectedErrMsg))
-      }
+      assert(toCharResult == dataTypeMismatch)
     }
   }
 
+  test("ToCharacter: fails analysis if numberFormat is not a constant") {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1127]: correct the title, it is not foldable

**Comment by MaxGekk** [comment]
also cc @dtenedor @cloud-fan @srielau @itholic

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (srielau) [reply core/src/main/resources/error/error-classes.json:295]: Seems there is 5 (or 3) of these, how about using suberror classes here:
INVALID_NUMBER_FORMAT.CURRENCY/EMPTY/SEPARATOR

Or:
INVALID_NUMBER_FORMAT_CURRENT
...

Also I think the two flavors for SEPARATOR and CURRENCY each can be merged.

Currency characters must appear before the decimal point and any digits in the number format

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:295]: Here are all subclasses of DATATYPE_MISMATCH.
Do we really need Level 3 error classes? @MaxGekk 

If we keep Level 2 error classes, INVALID_NUMBER_FORMAT....
The naming rule is inconsistent with other subclasses, eg:
RANGE_FRAME_INVALID_TYPE, RANGE_FRAME_MULTI_ORDER, RANGE_FRAME_WITHOUT_ORDER...

If we merge two flavors for SEPARATOR and CURRENCY, We have to put these characters into the message parameters, which seems inconvenient for subsequent internationalization.

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:295]: > Do we really need Level 3 error classes? @MaxGekk

I don't think it makes sense.

> If we keep Level 2 error classes, INVALID_NUMBER_FORMAT....

Let's make the prefix shorter, for instance `NUM_FORMAT` like
`DATATYPE_MISMATCH.NUM_FORMAT_CONT_THOUSANDS_SEPS`
 instead of
`DATATYPE_MISMATCH.NUMBER_FORMAT_THOUSANDS_SEPARATOR_MUST_HAVE_DIGIT_IN_BETWEEN`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 7 above)
Comment (panbingkun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1127]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 6 above)
Comment (panbingkun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1184]: Done

File: core/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:313]: Change it to 'NUMBER_FORMAT_SEPARATOR_MUST_HAVE_DIGIT_BETWEEN'?
It seems that the meaning of this error cannot be well displayed.

File: core/src/main/resources/error/error-classes.json (hunk 4 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:310]: Done

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:305]: Done

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:300]: Done

File: core/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -290,6 +290,46 @@
...
+        "message" : [
+          "Thousands separators (, or G) may not appear after the decimal point in the number format: <format>."
+        ]
+      },
+      "NUM_FORMAT_UNEXPECTED_TOKEN" : {
+        "message" : [
+          "Unexpected <token> found in the format string <format>; the structure of the format string must match: [MI|S] [$] [0|9|G|,]* [.|D] [0|9]* [$] [PR|MI|S]."
+        ]
+      },
+      "NUM_FORMAT_WRONG_NUM_DIGIT" : {
```
[the remark this answers is not in the document]
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Ok, let me do it.

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -290,6 +290,46 @@
           "Null typed values cannot be used as arguments of <functionName>."
         ]
       },
+      "NUM_FORMAT_CONT_THOUSANDS_SEPS" : {
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json]: How is this related to DATA_TYPE_MISMATCH?
Shouldn't this be a separate error_class:
INVALID_NUMBER_FORMAT.*

**Review by srielau — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 9 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json]: Theoretically, a data type is a set of possible values and a set of allowed operations on it. The number format is not an arbitrary string, it must be in some particular format. So, the error indicates violation of datatype range of values.

File: core/src/main/resources/error/error-classes.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done

File: core/src/main/resources/error/error-classes.json (hunk 9 above)
Comment (cloud-fan) [reply core/src/main/resources/error/error-classes.json]: I think we abuse type check in a few expressions... Some are query compile time static checks.

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json]: > Some are query compile time static checks

I would like to propose to do refactoring later. Let's assign unique error classes to the errors that come from `checkInputDataTypes()`.

**Review by MaxGekk — APPROVED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 9 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json]: @cloud-fan Are you ok with such approach?

Comment (srielau) [reply core/src/main/resources/error/error-classes.json]: This must be fixed before we cut 12.0.
Also there is no joy in having an error class with dozens of subclasses because it makes for a deeper doc tree.

By this logic a CONSTRAINT violation is a data type mismatch and so is a NOT NULL violation....

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json]: > This must be fixed before we cut 12.0.

Definitely, it will be fixed before Spark 12.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @cloud-fan @srielau for review.
